rollover.py

```diff
@@ -0,0 +1,300 @@
+#! /usr/bin/env python
+
+import argparse
+import itertools
+from operator import itemgetter
+import os
+import sys
+
+# local imports
+import ec2
+import elb
+import ecs
+import scaling
+import ssh
+import utils
+
+
+def sort_and_check_availability_zones(instance_descriptions, asg_instances):
+    """
+    sorts the instances into an order that tries not to cause an AZ imbalance
+    when removing instances. Also, prompts the user if there are issues.
+    @param instance_descriptions: dict of ec2_ids to descriptions
+    @param asg_instances: list of asg instance dicts
+    @return: sorted list of instances to remove
+    """
+    asg_instances_to_zones = {}
+    for asg_instance in asg_instances:
+        ec2_id = asg_instance['InstanceId']
+        asg_instances_to_zones[ec2_id] = asg_instance['AvailabilityZone']
+
+    # remove the selected instances to determine the remaining AZ balance
+    to_remove = {}
+    for ec2_id, desc in instance_descriptions.items():
+        if ec2_id in asg_instances_to_zones:
+            del asg_instances_to_zones[ec2_id]
+        else:
+            print "WARNING: %s is not in the AutoScalingGroup. It will not be replaced" % (ec2_id)
+
+        az = desc['Placement']['AvailabilityZone']
+        to_remove.setdefault(az, [])
+        to_remove[az].append(ec2_id)
+
+    # check az balance
+    max_diff = 0
+    for a, b in itertools.combinations(to_remove.keys(), 2):
+        max_diff = max(max_diff, abs(len(to_remove[a]) - len(to_remove[b])))
+
+    remaining = sum([len(i) for i in to_remove.values()])
+    ordered_instances = []
+    # order by zone with the most instances first
+    zone_counts = sorted(to_remove.items(),
+                         key=itemgetter(1),
+                         cmp=lambda a,b: cmp(len(a), len(b)),
+                         reverse=True)
+    for az in itertools.cycle([z[0] for z in zone_counts]):
+        if to_remove[az]:
+            ordered_instances.append(to_remove[az].pop(0))
+            remaining -= 1
+        if remaining == 0:
+            break
+
+    print "About to remove the following instances:"
+    for instance in ordered_instances:
+        print "%s - %s" % (instance, instance_descriptions[instance]['Placement']['AvailabilityZone'])
+
+    asg_zones = set([az for az in asg_instances_to_zones.values()])
+    if max_diff > 1 or len(asg_zones) == 1:
+        print "WARNING: The instances you selected will cause the auto scaling" \
+              " group to rebalance instances across availability zones. This" \
+              " will result in a destructive operation."
+
+    confirm = raw_input("Do you want to continue [y/N]? ")
+    if confirm.lower() != 'y':
+        return []
+
+    return ordered_instances
```

Comment:
this method seems unit testable, minus the `confirm` logic -- I'd consider moving that to the calling main method